File: tools/web.ts
export const webFetch: MarieTool = {
  name: 'web_fetch',
  description: 'Fetch the text content of a URL (HTML, JSON, plain text). Use this if the standard search is not enough.',
  safe: true,
  parameters: {
    type: 'object',
    properties: {
      url: { type: 'string', description: 'Full URL to fetch' },
    },
    required: ['url'],
  },
  run({ url }) {
    try {
      const { spawnSync } = require("bun");
      const proc = spawnSync(["curl", "-sL", "-A", "marie-universal/1", url as string]);
      if (proc.success) {
        return proc.stdout.toString().slice(0, 15000);
      }
      return `Error: curl failed with code ${proc.exitCode}`;
    } catch (e) {
      return `Error: ${e.message}`;
    }
  },
};

export const googleSearch: MarieTool = {
  name: 'google_search',
  description: 'Search the web for real-time information. A robust JS-based search tool.',
  safe: true,
  parameters: {
    type: 'object',
    properties: {
      query: { type: 'string', description: 'The search query' }
    },
    required: ['query']
  },
  run({ query }) {
    try {
        const { spawnSync } = require("bun");
        // Using DuckDuckGo Lite as a robust search backend for the JS bridge
        const searchUrl = `https://lite.duckduckgo.com/lite/?q=${encodeURIComponent(query as string)}`;
        const proc = spawnSync(["curl", "-sL", "-A", "Mozilla/5.0", searchUrl]);
        if (proc.success) {
          const html = proc.stdout.toString();
          // Extract result snippets (very basic parsing for the demo)
          const results = html.match(/<td class="result-snippet">[\s\S]*?<\/td>/g);
          if (results) {
              return results.slice(0, 5).map(r => r.replace(/<[^>]*>/g, '').trim()).join("\n---\n");
          }
          return "No search results found for this query.";
        }
        return `Error: search failed with code ${proc.exitCode}`;
    } catch (e) {
        return `Error: ${e.message}`;
    }
  }
};

File: clients/python/marie/tools/web.py
import requests
from .base import Tool

class WebFetchTool(Tool):
    def __init__(self):
        super().__init__(
            name="web_fetch",
            description="Fetch the content of a URL and return as text (max 5000 chars).",
            parameters={
                "type": "object",
                "properties": {
                    "url": {"type": "string", "description": "The URL to fetch"}
                },
                "required": ["url"]
            },
            safe=True
        )

    def run(self, url):
        try:
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            text = response.text
            if len(text) > 5000:
                text = text[:5000] + "... (truncated)"
            return text
        except Exception as e:
            return f"Error fetching URL: {str(e)}"
